org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ExtendedDocumentSymbolHandler.java

```diff
@@ -0,0 +1,186 @@
+import static org.eclipse.jdt.ls.core.internal.hover.JavaElementLabels.M_POST_QUALIFIED;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IParent;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeHierarchy;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+import org.eclipse.lsp4j.DocumentSymbol;
+import org.eclipse.lsp4j.DocumentSymbolParams;
+import org.eclipse.lsp4j.SymbolTag;
+import org.eclipse.lsp4j.jsonrpc.validation.NonNull;
+import org.eclipse.xtext.xbase.lib.Exceptions;
+import org.eclipse.xtext.xbase.lib.Pure;
+
+public class ExtendedDocumentSymbolHandler {
+	private PreferenceManager preferenceManager;
+
+	public ExtendedDocumentSymbolHandler(PreferenceManager preferenceManager) {
+		this.preferenceManager = preferenceManager;
+	}
+
+	public List<ExtendedDocumentSymbol> handle(DocumentSymbolParams params, IProgressMonitor monitor) {
+		ITypeRoot unit = JDTUtils.resolveTypeRoot(params.getTextDocument().getUri());
+		if (unit == null || !unit.exists()) {
+			return Collections.emptyList();
+		}
+		return getHierarchicalOutline(unit, monitor);
+	}
+
+	private List<ExtendedDocumentSymbol> getHierarchicalOutline(ITypeRoot unit, IProgressMonitor monitor) {
+		try {
+			return Stream.of(getInheritedChildren(unit, monitor)).map(child -> toDocumentSymbol(child, monitor)).filter(Objects::nonNull).collect(Collectors.toList());
+		} catch (OperationCanceledException e) {
+			logInfo("User abort while collecting the document symbols.");
+		} catch (JavaModelException e) {
+			if (!unit.exists()) {
+				JavaLanguageServerPlugin.logError("Problem getting outline for " + unit.getElementName() + ": File not found.");
+			} else {
+				JavaLanguageServerPlugin.logException("Problem getting outline for " + unit.getElementName(), e);
+			}
+		}
+		return emptyList();
+	}
+
+	private static IJavaElement[] getInheritedChildren(ITypeRoot unit, IProgressMonitor monitor) throws JavaModelException {
+		IType type = unit.findPrimaryType();
+		List<IJavaElement> children = new ArrayList<>();
+		if (type != null) {
+			children.addAll(Arrays.asList(filter(type.getChildren())));
+			if (monitor.isCanceled()) {
+				throw new OperationCanceledException();
+			}
+			ITypeHierarchy supertypeHierarchy = type.newSupertypeHierarchy(monitor);
+			IType[] subclasses = supertypeHierarchy.getAllSuperclasses(type);
+			for (IType subType : subclasses) {
+				if (monitor.isCanceled()) {
```

Comment:
The for loop here looks not very expensive. Instead of check cancel in each loop, I would recommend checking it after a method invocation which consume a monitor. For example, after `type.newSupertypeHierarchy(monitor);`